caffe2/operators/pool_op_util.cc

```diff
@@ -0,0 +1,922 @@
+    const int kw,
+    const int stride_h,
+    const int stride_w,
+    const int pad_t,
+    const int pad_l,
+    const int pad_b,
+    const int pad_r,
+    const int dilation_h,
+    const int dilation_w,
+    const float* X,
+    float* Y) {
+#if defined(__ARM_NEON__) || defined(__ARM_NEON)
+  // Use this kernel only if:
+  //   1. Kernel size is 2x2
+  //   2. Stride is 2x2
+  //   3. Padding is 0
+  //   4. Dilation is 1
+  //   5. Output width and height are even divisors of input width
+  //   6. Input width and height are divisible by 4 (should be implied b all of
+  //      the above, but just check again)
+  // Input and output pointers are aligned by float32x4_t
+  const bool kernel_ok = (kh == 2) && (kw == 2);
+  const bool stride_ok = (stride_h == 2) && (stride_w == 2);
+  const bool pad_ok =
+      (pad_t == 0) && (pad_l == 0) && (pad_b == 0) && (pad_r == 0);
+  const bool dilation_ok = (dilation_h == 1) && (dilation_w == 1);
+  const bool output_ok = (input_h % output_h == 0) && (input_w % output_w == 0);
+  const bool input_ok = (input_w % 4 == 0) && (input_h % 4 == 0);
+  const bool align_ok = isPointerAligned(X, sizeof(float32x4_t)) &&
+      isPointerAligned(Y, sizeof(float32x4_t));
+  return kernel_ok && stride_ok && pad_ok && dilation_ok && output_ok &&
+      input_ok && align_ok;
+#else
+  (void)input_h;
+  (void)input_w;
+  (void)output_h;
+  (void)output_w;
+  (void)kh;
+  (void)kw;
+  (void)stride_h;
+  (void)stride_w;
+  (void)pad_t;
+  (void)pad_l;
+  (void)pad_b;
+  (void)pad_r;
+  (void)dilation_h;
+  (void)dilation_w;
+  (void)X;
+  (void)Y;
+  return false;
+#endif
+}
+
+void RunNeonAveragePool4x4p0s0NCHW(
+    int N,
+    int C,
+    int H,
+    int W,
+    const float* X,
+    float* Y) {
+#if defined(__ARM_NEON__) || defined(__ARM_NEON)
+  const int X_stride = H * W;
+  const int Y_stride = (H / 4) * (W / 4);
+  const float* X_ptr = X;
+  float* Y_ptr = Y;
+  for (int i = 0; i < N; ++i) {
+    for (int j = 0; j < C; ++j) {
+      AvgPoolNeon4x4p0s0Plane(H, W, X_ptr, Y_ptr);
+      X_ptr += X_stride;
+      Y_ptr += Y_stride;
+    }
+  }
+#else
+  (void)N;
+  (void)C;
+  (void)H;
+  (void)W;
+  (void)X;
+  (void)Y;
+#endif
+}
+
+void RunNeonMaxPool2x2p0s0NCHW(
+    int N,
+    int C,
+    int H,
+    int W,
+    const float* X,
+    float* Y) {
+#if defined(__ARM_NEON__) || defined(__ARM_NEON)
+  const int X_stride = H * W;
+  const int Y_stride = (H / 2) * (W / 2);
+  const float* X_ptr = X;
+  float* Y_ptr = Y;
+  for (int i = 0; i < N; ++i) {
+    for (int j = 0; j < C; ++j) {
+      MaxPoolNeon2x2p0s0Plane(H, W, X_ptr, Y_ptr);
+      X_ptr += X_stride;
+      Y_ptr += Y_stride;
+    }
+  }
+#else
+  (void)N;
+  (void)C;
+  (void)H;
+  (void)W;
+  (void)X;
+  (void)Y;
+#endif
+}
+
+#define CAFFE2_SPECIALIZED_RUN_AVERAGE_POOL_1D(T, kCountIncludePad)    \
+  template <>                                                          \
+  void RunAveragePool1D<T, StorageOrder::NCHW, kCountIncludePad>(      \
```

Comment:
Can we add kCountIncludePad as argument?

Reply:
It is trying to make the design of the related functions on CPU and GPU the same. On GPU part, if there is a condition between different threads in runtime, there will be some negative impact on the performance.